grails-app/services/org/pih/warehouse/invoice/PrepaymentInvoiceService.groovy

```diff
@@ -352,4 +366,78 @@ class PrepaymentInvoiceService {
 
         return quantityInvoiced >= quantityInverseable ? quantityInverseable : quantityInvoiced
     }
+
+    private void validateItem(InvoiceItem invoiceItem, String itemId) {
+        if (!invoiceItem) {
+            String defaultMessage = "Cannot find invoice item with id: ${itemId}"
+            throw new IllegalArgumentException(applicationTagLib.message(code: "invoiceItem.cannotFind.error", args: [itemId], default: defaultMessage))
+        }
+        if (invoiceItem.invoice.datePosted) {
+            String defaultMessage = "Cannot update posted invoices"
+            throw new IllegalArgumentException(applicationTagLib.message(code: "invoice.cannotUpdatePosted.error", default: defaultMessage))
+        }
+        if (invoiceItem?.inverse) {
+            String defaultMessage = "Cannot edit inverse invoice items directly"
+            throw new IllegalArgumentException(applicationTagLib.message(code: "invoiceItem.cannotEditInverse.error", default: defaultMessage))
+        }
+    }
+
+    /**
+     * Updates invoice item's unit price and amount and finds related inverse item and updates unit price and amount
+     * */
+    void updateInvoiceItemUnitPrice(String itemId, BigDecimal unitPrice) {
+        InvoiceItem invoiceItem = InvoiceItem.get(itemId)
+        validateItem(invoiceItem, itemId)
+        if (unitPrice == 0) {
+            String defaultMessage = "Cannot update unit price to 0"
+            throw new IllegalArgumentException(applicationTagLib.message(code: "invoiceItem.unitPriceZero.error", default: defaultMessage))
+        }
+        if (!invoiceItem?.orderAdjustment) {
+            String defaultMessage = "Invoice item is missing order adjustment. Cannot edit unit price."
+            throw new IllegalArgumentException(applicationTagLib.message(code: "invoiceItem.missingAdjustment.error", default: defaultMessage))
+        }
+        BigDecimal totalAdjustments = invoiceItem.orderAdjustment?.totalAdjustments
+        if ((totalAdjustments < 0 && unitPrice > 0) || (totalAdjustments > 0 && unitPrice < 0)) {
+            String defaultMessage = "Cannot change the positive unit price to negative or negative to positive."
+            throw new IllegalArgumentException(applicationTagLib.message(code: "invoiceItem.unitPriceSignChange.error", default: defaultMessage))
+        }
+
+        BigDecimal amountAvailableToInvoice = invoiceItem.orderAdjustment.amountAvailableToInvoice
+        if (Math.abs(unitPrice) > Math.abs(amountAvailableToInvoice) + Math.abs(invoiceItem.unitPrice)) {
+            String defaultMessage = "Cannot update unit price to higher value than available to invoice"
+            throw new IllegalArgumentException(applicationTagLib.message(code: "invoiceItem.unitPriceTooHigh.error", default: defaultMessage))
+        }
```

Comment:
is there any reason why those blocks can't also be placed in the `validateItem` method?

Reply:
moreover I think those should not throw `IllegalArgumentException` as this would lead to 500 being thrown, instead of 400 that are expected for validation errors.
This leads me to a statement that this would be probably better to place this type of validation in an appropriate command object and a solid validator. Would be easier to maintain this logic for any property, and also to maintain the error messages.

Reply:
> is there any reason why those blocks can't also be placed in the validateItem method?

I just moved out the common part from `updateInvoiceItemQuantity` and `updateInvoiceItemUnitPrice` and remaining ones are different for quantity and unit price update cases. 

> moreover I think those should not throw IllegalArgumentException 

Yup, I agree with the ValidationException being more suitable here, but yesterday I thought that change could be done together with adding a / moving to command in a separate ticket, that's why I left these. I'll check if simple exception type change will work here (if it is a quick change I'll swap it).

Reply:
@awalkowiak unfortunately it won't be a simple change. Since you don't bind the errors to the object, and we were to throw the validation exception, the object's errors would be empty.
If we wanted to do it via your methods, you would have to do something like:
```
invoiceItem.errors.rejectValue(....)
```
but I agree and I am fine doing it in the cleanup ticket, we can build a command there, and construct a proper validator.

Reply:
We just need to be aware not to complicate the frontend error handling too much, as the response status would eventually change from 500 to 400 when we do the cleanup, and then, the axios error handler might build the popup differently.
cc @alannadolny @drodzewicz

Reply:
relevant to this discussion, I have a ticket for revisiting how we handle validation errors on the backend https://pihemr.atlassian.net/browse/OBPIH-6727

I hope that after that standard is built, a lot of this type of validation can be simplified.